Add function to include xUnit.NET tests

fmrico opened the pull request:
Hi @esteve 

PR esteve/ros2_dotnet#10 contributes to the opportunity of performing xUnit tests. Furthermore, each project could require a different .NET standard, as happens with xUnit.

In order to include more than one C# project in a single ros2 package, it is necessary to do some changes in this repo/package. The proposed changes let to build individually several C# projects in the same package. 

Best

fmrico committed:
Add function to include xUnit.NET tests

esteve commented:
Thanks for the PR. Correct me if I'm wrong, but it seems that this won't add the test to CMake, it just compiles the source code and requires the user to type `dotnet test`. If so, `dotnet test` already compiles the code before running the tests, so the function might not add much.

Just a side note, could you describe with more detail in your PRs what the changes are and what the motivation is? It helps a lot to understand the changes.

fmrico commented:
Actually, I still don't know how to integrate it with cmake. Could you please give me a pointer?

Yes, the description of the PR is very short. I'll try to fix as soon as posible.

fmrico committed:
Several C# projects in same package; Test integration with ament

Signed-off-by: Francisco Martín Rico <fmrico@gmail.com>

esteve commented on cmake/Modules/dotnet/UseCSharpProjectBuilder.cmake:
```diff
@@ -177,6 +178,14 @@ function(csharp_add_project name)
         DEPENDS ${sources_dep}
     )
 
+    if (${_csharp_add_project_TEST_DLL})
```
This should be part of `FindDotNETExtra.cmake`, none of the code in `UseCSharpProjectBuilder.cmake` is dotnet-specific

fmrico replied:
Ok. Changed

esteve commented on cmake/Modules/FindDotNETExtra.cmake:
```diff
@@ -58,6 +58,32 @@ function(add_dotnet_executable _TARGET_NAME)
+    ${_add_dotnet_test_INCLUDE_NUPKGS}
+    INCLUDE_REFERENCES
+    ${_add_dotnet_test_INCLUDE_REFERENCES}
+    TEST_DLL
```
The xUnit assemblies should be added here.

fmrico replied:
Ok. Added

esteve reviewed (changes requested):
It's looking good, there's just a couple of changes and this should be ready to go. Please, update the description with more details, it's easier to track what a PR does.

esteve commented on cmake/Modules/dotnet/Directory.Build.props.in:
```diff
@@ -0,0 +1,5 @@
+<Project>
```
Could you explain the rationale for this change?

fmrico replied:
The directory for building each of the projects must be independent. The best option I have found makes it depends on his project name.

esteve replied:
Ok, but why does it need to be independent? What's the rationale?

fmrico replied:
Ok. If my memory doesn't fail me, xunit required a different target framework. As buildings were made in the same directory, concurrent buildings that required different target frameworks failed. 

This change lets to build different project in different build directories to avoid this problem.

fmrico commented:
I have just detected a race condition during building. Stay tuned for the fix...

fmrico committed:
Fix concurrency problem when building several projects

Signed-off-by: Francisco Martín Rico <fmrico@gmail.com>

fmrico commented:
Hi @esteve ,

I have just fixed a concurrency problem that happens when building several projects in the same package when they require a different .NET standard (as happens when building xUnit tests). It is caused by the concurrent access to the intermediate files used when building.

I have fixed this problem by building projects in different directories, each one with the name of the project.

Best

esteve commented:
@fmrico thanks for looking into the concurrency bug. Do you think you'll have a moment to address the rest of the feedback and to update the description of the pull request? Thanks.

fmrico committed:
Moving dotnet-specific cmake code to FindDotNETExtra.cmake

Signed-off-by: Francisco Martín Rico <fmrico@gmail.com>

fmrico committed:
Include xUnit references in add_dotnet_test function

Signed-off-by: Francisco Martín Rico <fmrico@gmail.com>

fmrico commented:
@esteve I think I have just resolved the question you requested and I improved this PR description.

fmrico commented:
@esteve ping

esteve commented:
@fmrico do these changes actually report the tests that fail/pass? `ament_add_test` needs an output file to read result from and `dotnet test` doesn't seem to generate one.

esteve commented:
@fmrico any updates?

fmrico commented:
Hi @esteve,

No updates. Right now we are facing a connection problem in uwp32. We are debugging if this a problem related to any update in Windows 10.

I'll come back on this as soon as possible,

Best

esteve commented:
@fmrico I meant if you have any updates on the the feedback I gave you (output file for ament_add_test, explaining the reason for splitting the csproj file, etc.)

fmrico commented:
Hi @esteve 

Let's retake this PR.

fmrico commented:
@esteve ping

theseankelly commented on cmake/Modules/FindDotNETExtra.cmake:
```diff
@@ -58,6 +58,50 @@ function(add_dotnet_executable _TARGET_NAME)
+    ${ARGN}
+  )
+
+  set(CSHARP_TARGET_FRAMEWORK "netcoreapp2.0")
```
CSHARP_TARGET_FRAMEWORK is set by the file `FindCSBuild.cmake` -- were you explicit here for unit tests on purpose, or should the same dotnet target framework be used as what the rest of the project was built with?

fmrico replied:
Using netstandard1.1 with xUnit, I got a `"CS5001 error: Program does not contain a static 'Main' method suitable for an entry point"`. I couldn't resolve this problem, as every example that I found used netcoreapp2.0. 

CSHARP_TARGET_FRAMEWORK is set by default to "netstandard1.1 in the file FindCSBuild.cmake:34. I have changed to "netcoreapp2.0" for being used as default in rcldotnet at everything looks working ok. I only tested in Linux. 

In [this](https://docs.microsoft.com/en-us/archive/msdn-magazine/2016/april/net-core-net-goes-cross-platform-with-net-core) article says that UWP apps can only reference the .NET standard class library template, not .NET Core class library template, so using "netcoreapp2.0" as default is not an option by default.

Summarizing: We need "netcoreapp2.0" for testing with XUnit and "netstandard1.1" for the rest of the project if we want to run UWP.

theseankelly replied:
Makes sense -- thanks for taking the time to help me understand.

theseankelly commented on cmake/Modules/dotnet/UseCSharpProjectBuilder.cmake:
```diff
@@ -154,26 +154,27 @@ function(csharp_add_project name)
-        -DCONFIG_OUTPUT_FILE="${CURRENT_TARGET_BINARY_DIR}/Directory.Build.props"
+        -DCSHARP_BUILDER_OUTPUT_NAME="${name}${CSBUILD_OUTPUT_SUFFIX}"
+        -DCONFIG_INPUT_FILE="${dotnet_cmake_module_DIR}/Modules/dotnet/Directory.Build.props.in"
+        -DCONFIG_OUTPUT_FILE="${CURRENT_TARGET_BINARY_DIR}/${name}/${name}$/Directory.Build.props"
```
This feels like a typo based on the pattern in the rest of the changes, but I haven't thought/tested in great detail.

Did you mean to have ${name}/${name}$/Directory.Build.props?

theseankelly replied:
Been playing around with the change -- I do think this is a typo. Resulting build directory:

build\rcldotnet_examples\rcldotnet_listener:
![image](https://user-images.githubusercontent.com/1154591/80304306-ec68a280-8782-11ea-9d98-ade7c46572e7.png)

fmrico replied:
Exactly. This is a typo

theseankelly replied:
Pardon me being pedantic, but just confirming the duplicate name (`${name}/${name}`) is intentional? Or should the Directory.Build.props file be in the root along with the csproj?

Again I don't know this stuff too well; I only ask because everywhere else you've only added one level of subdirectory.

fmrico replied:
I have changed it, and everything is working, and tests are passing. I thought that it was because of some concurrency problems that I had while building. I have removed it, and I hope not to revert this change in the future...

theseankelly replied:
If you think there was a reason for it, I believe you and go ahead and leave it.

Again, my comments are mostly based on pattern matching in the diffs, not deep analysis of the build tree!

theseankelly reviewed (changes requested):
Sorry for my very late review; it's been that kind of a month.

Just a couple points/questions, mostly for my own education

fmrico committed:
Fixed typo in path

Signed-off-by: Francisco Martin Rico <fmrico@gmail.com>

fmrico committed:
Removing duplicate nested dir

Signed-off-by: Francisco Martin Rico <fmrico@gmail.com>
